### PR TITLE
Decomposition of aten.pixel_shuffle with static input shape

### PR DESCRIPTION
**Pull request by newling**
For static tests (that is when the shape is know) for example:

 ```
 @annotate_args([None, ([3, 18, 2, 2], torch.float32, True)])
 ```
 
 The e2e passes. But only if the replacement op's return type is set as undefined (optional shape and type must be explicitly made unset), otherwise there's a error about the function return type. 
 
 For dynamic cases, for example if the above is replaced with 
 
  ```
 @annotate_args([None, ([-1, -1, -1, -1], torch.float32, True)])
 ```

There is a failure to lower to linalg from torch ("view op explicitly labelled as illegal"). This seems to be because the support for lowering from torch to linalg with dynamic shapes is limited.

### CODE REVIEW

**Comment by newling** [comment]
@vivekkhandelwal1 I have an e2e test case working, but only for static shapes (see my comment above). I am not sure what to do about the dynamic shape case, is my thought that torch->linalg for aten.view is limited, and that's why I'm having problems? 

More generally, I'd like to understand dynamic shapes end-to-end in torch-mlir, any pointers?

**Comment by vivekkhandelwal1** [comment]
> @vivekkhandelwal1 I have an e2e test case working, but only for static shapes (see my comment above). I am not sure what to do about the dynamic shape case, is my thought that torch->linalg for aten.view is limited, and that's why I'm having problems?
> 
> More generally, I'd like to understand dynamic shapes end-to-end in torch-mlir, any pointers?

Hi @newling, the support for dynamic shapes in Torch-MLIR is limited, especially for view-like ops. I think, for now, you can follow either of the below two paths:
1.) Add a condition in your decomp restricting the support for only static shapes (Easy and fast path, but limited support).
2.) Directly add a `Torch -> Linalg` lowering for this op, if it seems feasible (Hard and slow path, but you will get complete support for the op).

For the `Torch -> Linalg` support you can refer to this: https://github.com/pytorch/pytorch/blob/fa9045a8725214c05ae4dcec5a855820b861155e/aten/src/ATen/native/cpu/PixelShuffleKernel.cpp#L16-L53

Also, unfortunately, there is no documentation for the dynamic shapes in Torch-MLIR. It's just that ops in Torch-MLIR broadly fall into one or a combination of the below four categories:
1.) Full-Dynamic Support
2.) Partial-Dynamic Support
3.) Full-Static Support
4.) Partial-Static Support.

You can go through some of the op lowerings for view-like ops to understand the functioning or dynamic-shape support in Torch-MLIR.

**Comment by newling** [comment]
Thanks for these information gems @vivekkhandelwal1 

I am tempted to go with option 1 (just support static) for now, and leave comments in the code about what the issues currently are with dynamic shapes, and what possible remedies there are (like go directly to linalg, or maybe extend lowering of reshape). 

Hopefully the use cases that @qedawkins  has in mind are static.

**Comment by vivekkhandelwal1** [comment]
> Thanks for these information gems @vivekkhandelwal1
> 
> I am tempted to go with option 1 (just support static) for now, and leave comments in the code about what the issues currently are with dynamic shapes, and what possible remedies there are (like go directly to linalg, or maybe extend lowering of reshape).
> 
> Hopefully the use cases that @qedawkins has in mind are static.

If it's required for Llama-2 then the use case would probably be dynamic. But @qedawkins can comment more on this,

**Comment by qedawkins** [comment]
> > Thanks for these information gems @vivekkhandelwal1
> > I am tempted to go with option 1 (just support static) for now, and leave comments in the code about what the issues currently are with dynamic shapes, and what possible remedies there are (like go directly to linalg, or maybe extend lowering of reshape).
> > Hopefully the use cases that @qedawkins has in mind are static.
> 
> If it's required for Llama-2 then the use case would probably be dynamic. But @qedawkins can comment more on this,

Ah no, this mostly shows up in vision models, and I would guess most of the time we would see static dims for the ones that are actually shuffled. There might be some dynamic outer/batch like dims though.

**Comment by newling** [comment]
Ok, so I've got it supporting just static shapes. I've left a comment in the code about why, and possible next steps. 

I don't seem to be able to add reviewers, and I also need approval to start the workflows. How I get these permissions?

**Comment by qedawkins** [comment]
> Ok, so I've got it supporting just static shapes. I've left a comment in the code about why, and possible next steps.
> 
> I don't seem to be able to add reviewers, and I also need approval to start the workflows. How I get these permissions?

Good question. IIRC there is a process to be added to the repo after you have landed a commit, but I don't know if the process has changed.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1156]: There are likely other benefits of doing this. In particular, the kind of reshape done by this operation is relatively structured. This makes the analysis needed on the reshape ops to recover the optimal expand/collapse_shape ops much easier (especially in dynamic cases).

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1163]: Torch-mlir cannot decompose to a tensor operation. One important part of torch-mlir is the ability to lower to different target IRs (or even mix-and-match), thus all torch decompositions must be strictly in terms of torch ops.

In regards to `builtin.unrealized_conversion_cast` though, there is a `torchc.to_builtin` and `torchc.from_builtin` (might have gotten the names wrong) that allow conversion between the builtin tensor type and the torch tensor type.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1168]: It would definitely interesting to have something along the lines of a `TorchExt` extension of the torch dialect (or just a set of add on operations) that torch operators can decompose to. This could help unify lowerings for particularly complicated ops by first decomposing them a bit if there is no good choice already in ATen.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1136]: There is a WIP patch to help with these unhandled dynamic cases (requires some upstream changes to tensor.expand_shape), but it will still generate somewhat non-optimal IR by essentially doing a full collapse -> full expand.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1295]: Can you elaborate a bit on what was going wrong here?

**Review by qedawkins — COMMENTED** [review]
Mostly looks good, thanks for the patch! It has been a while since I've worked in depth with these decompositions and there might be some helpers already in place to help simplify code in a few places (looking for another reviewer who might know).

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:232]: (nothing actionable here, just a note)

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1156]: Ah, I read the earlier part of the comment more closely. Yes, +1 to this approach.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1163]: Makes sense, so this isn't an option because we it'd break the lowering to tosa, for example. Good to know that about the to_builtin and from_builtin.  Out of curiosity, do you know why torch-mlir decided against using the builtin tensor type to represent a value semantics tensor?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1168]: I'd be happy to explore this as a follow-up task if other reviewers don't see any problem with this approach. One concern I have is that introducing a new "concept" to torch-mlir like this might make the lowering paths less "obvious".

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1136]: Yeah. In a way it is good that the full collapse path isn't in torch-mlir yet -- if it was this PR would have just worked with dynamic shapes and we would have missed the possible better loop nesting.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1181]: I'll update the comment in the code as:

The input tensor must have at least 3 dimensions: (1) the channel dimension which gets smaller by 'factor*factor', (2) the H channel which gets larger by 'factor' and (3) the W channel which get larger by 'factor'. The total number of dimensions is 3 + N, where N is the number of leading dimensions (N >= 0).

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1295]: Below, "finalShape" is  `!torch.vtensor<[3,4,6],si64>`.

But I think "op" has type `!torch.vtensor` (?), and so if I replace 

` rewriter.replaceOpWithNewOp<AtenReshapeOp>(op, finalType, B, outShapeList);`

in the code with 

`rewriter.replaceOpWithNewOp<AtenReshapeOp>(op, finalShape, B, outShapeList);`

I get the error:

```
        python exception: Failure while executing pass pipeline:
        error: unknown: type of return operand 0 ('!torch.vtensor<[3,4,6],si64>') doesn't match function result type ('!torch.vtensor') in function @forward
        note: unknown: see current operation: "func.return"(%11) : (!torch.vtensor<[3,4,6],si64>) -> ()
```

Sorry, my variable names "finalShape" and "finalType" are bad.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:232]: That makes sense, noted for next time 👍

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: (The leading dims can be empty)

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Ah, so the reason for this is that the rewrite that you do on this operation must replace all original uses of `aten.pixel_shuffle` with values of the same type. Like you said, the op has type `!torch.vtensor`, so we see something like this
```
%0 = torch.aten.pixel_shuffle ... -> !torch.vtensor
%1 = "some_user" %0 : !torch.vtensor -> !torch.int
```
If you don't use the same type, the resulting IR will look like this
```
%0 = torch.aten.pixel_shuffle ... -> !torch.vtensor<[2, 3, 6]xf32>
%1 = "some_user" %0 : !torch.vtensor -> !torch.int
```
and thus there is a type mismatch between %0 and it's user, %1. This is because all operations carry the types of their operands, and you can't change the type of an op operand on the fly.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1170]: Also, typically long TODO comments like this are best for github issues rather than inline in the code. Can you keep a short-form of `1)` as a TODO comment and move the rest to an issue? For example
```
// TODO: Lower pixel_shuffle to linalg directly. Pixel shuffle does a static reshape that is hard to recover through pure torch (view) ops, especially in dynamic cases.
```

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:12]: nit: What do you need to use from Tensor?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1300]: I think the reason you need to cast back to an unranked tensor is because you are missing a shape conversion function here: https://github.com/llvm/torch-mlir/blob/main/projects/pt1/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py

There should be an entry for `aten.pixel_shuffle` in `include/torch-mlir/Dialect/Torch/IR/JITOperatorRegistryDump.txt` that has shape/dtype/(decomposition) signature functions that you can add. Here is an example for aten._fw_primal
```
  shape_function_signature = def aten〇_fw_primal〡shape(self: List[int], level: int) -> List[int]:
  decomposition_function_signature = def aten〇_fw_primal〡decomposition(self: Tensor, level: int) -> Tensor:
  dtype_function_signature = def aten〇_fw_primal〡dtype(self_rank_dtype: Tuple[int, int], level: int) -> int:

```

**Review by qedawkins — CHANGES_REQUESTED** [review]
I left the comments that still need addressing as unresolved. Mainly some comment cleanup and missing shape/dtype refinement functions (sorry for the segmented review).

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (qedawkins) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: I think it was some attempt to reflect the more general programming model that pytorch offers, and since a non-value semantics tensor was going to be needed anyway, I'm guessing the idea was that swapping between builtin types and a torch type would make things more difficult. I wasn't around for those design decisions though :)

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:1300]: Nice, I was wondering where to implement the type inference. That fixes the problem.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp
Comment (newling) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:12]: Removed (was experimenting with unrealized_conversion to use expand_shape and collapse_shape)

**Review by qedawkins — APPROVED** [review]
LGTM! Thanks for pushing this through

**Comment by qedawkins** [comment]
I think some linting check wants you to run
```
./build_tools/update_abstract_interp_lib.sh
```

**Comment by newling** [comment]
Thanks for the help @qedawkins! 

> I think some linting check wants you to run
> 
> ```
> ./build_tools/update_abstract_interp_lib.sh
> ```

Yeah, that was a very useful diagnostic message from CI 😍

**Comment by newling** [comment]
@qedawkins would you mind kickstarting CI? It was failing with an "unexpected pass" failure, so I added the 2 tests to projects/pt1/e2e_testing/xfail_sets.py (a bit of a shot in the dark) 

General question: Is there a simple way to run all the tests locally, with a single command?

**Comment by qedawkins** [comment]
There are some notes on testing here: https://github.com/llvm/torch-mlir/blob/main/docs/development.md#testing
(I don't remember off the top of my head what the command you're looking for is)

**Comment by newling** [comment]
There isn't much material about updating xfail_sets.py... but I'm optimistic I've got it right now (added to LTC_XFAIL_SET). Can you please restart the workflow @qedawkins  ?

**Comment by qedawkins** [comment]
> There isn't much material about updating xfail_sets.py... but I'm optimistic I've got it right now (added to LTC_XFAIL_SET). Can you please restart the workflow @qedawkins ?

Sorry about needing approval each time to run the CI :(
Let's get this in and get you direct access

File: projects/pt1/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py
Comment (ramiro050) [inline projects/pt1/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py:427]: nit: all dtype functions here are tested e2e. Can you add a test to this one? It should be a matter of adding something like `@check_dtype_function(_check_tensors_with_the_same_dtype(num_of_tensors=1, upscale_factor=1))`

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py
Comment (ramiro050) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py:669]: nit: no need to write the shape in the test name. I think the important bit here is the input dtype, so I would call the tests something like `PixelShuffleModuleStaticInt64` and `PixelShuffleModuleStaticFloat32`

**Review by ramiro050 — COMMENTED** [review]
Some small requests

File: projects/pt1/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py
Comment (ramiro050) [reply projects/pt1/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py:427]: You can run the tests by generating the shape and dtype library: `./build_tools/update_abstract_interp_lib.sh`

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py
Comment (newling) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py:669]: Initially my motivation for the second test was for the rank-3 case (no leading dimensions). Making it a different dtype was just an extra, to get coverage in a different direction.  I've gone with names `Rank4Float32` and `Rank3Int64`

File: projects/pt1/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py
Comment (newling) [reply projects/pt1/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py:427]: Done!

**Comment by qedawkins** [comment]
(Merging because it looks like Ramiro's concerns were addressed and there is no PR blocker in place)

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py
Comment (ramiro050) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py]: Makes sense, thanks!

**Pull request merged**
